### PR TITLE
feat: expose hnsw indices

### LINKED ISSUE #1522: Feature: expose HNSW indices in Python and Node
### SDK

Python

### Description

HNSW indices are configurable in Rust but not in Python or Node yet.

There is support for several index types that aren't yet exposed:

https://github.com/lancedb/lancedb/blob/9c1adff4268fc739de36ca8d0f0dfebd2114b726/rust/lancedb/src/table.rs#L1680-L1704

See also: #1511

### PR DESCRIPTION
**Pull request by gagan-bhullar-tech**
PR closes #1522

### CODE REVIEW

**Comment by gagan-bhullar-tech** [comment]
@wjones127 can you please review PR. I will create a separate PR for python. Thanks.

**Comment by gagan-bhullar-tech** [comment]
@wjones127 can you please review PR.

File: python/src/index.rs [hunk 1]
```diff
@@ -105,6 +105,20 @@ impl Index {
...
+
+    #[staticmethod]
+    pub fn hnswpq() -> PyResult<Self> {
+        Ok(Self {
+            inner: Mutex::new(Some(LanceDbIndex::IvfHnswPq(Default::default()))),
+        })
+    }
+
+    #[staticmethod]
+    pub fn hnswsq() -> PyResult<Self> {
```
Comment (wjones127) [inline python/src/index.rs:117]: We should be able to set the parameters for these. See `ivf_pq` method above.

**Review by wjones127 — COMMENTED** [review]
@gagan-bhullar-tech HNSW indices have several configuration options such as `m`, `m_max`, `ef_construction`. These already exist here: https://github.com/lancedb/lancedb/commit/5e0181043845afd61c4de32e3ef05acef87f495b

**Comment by gagan-bhullar-tech** [comment]
@wjones127 @westonpace can you please review.

File: nodejs/lancedb/indices.ts [hunk 2]
```diff
@@ -214,6 +233,43 @@ export class Index {
   static fts() {
     return new Index(LanceDbIndex.fts());
   }
+
+  /**
+   *
+   * Create a hnswpq index
+   *
+   */
+  static hnswpq(options?: Partial<HnswPqOptions>) {
```
Comment (wjones127) [inline nodejs/lancedb/indices.ts:242]: To be consistent with other camel case names:
```suggestion
  static hnswPq(options?: Partial<HnswPqOptions>) {
```

File: nodejs/lancedb/indices.ts [hunk 3]
```diff
@@ -214,6 +233,43 @@ export class Index {
...
+      ),
+    );
+  }
+
+  /**
+   *
+   * Create a hnswsq index
+   *
+   */
+  static hnswsq(options?: Partial<HnswSqOptions>) {
```
Comment (wjones127) [inline nodejs/lancedb/indices.ts:261]:
```suggestion
  static hnswSq(options?: Partial<HnswSqOptions>) {
```

File: python/python/lancedb/index.py [hunk 4]
```diff
@@ -82,6 +82,54 @@ def __init__(self):
         self._inner = LanceDbIndex.fts()
 
 
+class Hnswpq:
```
Comment (wjones127) [inline python/python/lancedb/index.py:85]:
```suggestion
class HnswPq:
```

File: python/python/lancedb/index.py [hunk 5]
```diff
@@ -82,6 +82,54 @@ def __init__(self):
...
+            num_partitions=num_partitions,
+            num_sub_vectors=num_sub_vectors,
+            max_iterations=max_iterations,
+            sample_rate=sample_rate,
+            m=m,
+            ef_construction=ef_construction,
+        )
+
+
+class Hnswsq:
```
Comment (wjones127) [inline python/python/lancedb/index.py:110]:
```suggestion
class HnswSq:
```

File: python/python/lancedb/index.py [hunk 6]
```diff
@@ -82,6 +82,54 @@ def __init__(self):
...
+    """Describe a Hnswsq index configuration."""
```
Comment (wjones127) [inline python/python/lancedb/index.py:111]:
```suggestion
    """Describe a HNSW-SQ index configuration."""
```

**Review by wjones127 — CHANGES_REQUESTED** [review]
HNSW and PQ and SQ are each their own "word", so please capitalize the variables names appropriately. Be consistent with how we handle `IVF_PQ` indices.

It looks like you've added all the options. But they are not documented. I'm fine with leaving the documentation for a future PR.

**Comment by gagan-bhullar-tech** [comment]
@wjones127 please review

**Review by wjones127 — APPROVED** [review]
This is looking good now. I can fix the merge conflict in a minute.

**Comment by wjones127** [comment]
Added a follow up issue for documentation: https://github.com/lancedb/lancedb/issues/1627

**Pull request merged**